contracts/core/GlobalReentrancyLock.sol

```diff
@@ -0,0 +1,148 @@
+    // pointer aliasing, and it cannot be disabled.
+
+    // The values being non-zero value makes deployment a bit more expensive,
+    // but in exchange the refund on every call to nonReentrant will be lower in
+    // amount. Since refunds are capped to a percentage of the total
+    // transaction's gas, it is best to keep them low in cases like this one, to
+    // increase the likelihood of the full refund coming into effect.
+    uint8 private constant _NOT_ENTERED = 1;
+    uint8 private constant _ENTERED = 2;
+
+    /// -------------------------------------------------
+    /// -------------------------------------------------
+    /// -------------- Single Storage Slot --------------
+    /// -------------------------------------------------
+    /// -------------------------------------------------
+
+    /// @notice cache the address that locked the system
+    /// only this address can unlock it
+    uint160 private _sender;
+
+    /// @notice store the last block entered
+    /// if last block entered was in the past and status is entered, the system is in an invalid state
+    /// which means that actions should be allowed
+    uint88 private _lastBlockEntered;
+
+    /// @notice whether or not the system is entered or not entered
+    uint8 private _status;
+
+    /// @notice construct GlobalReentrancyLock
+    constructor() {
+        _status = _NOT_ENTERED;
+    }
+
+    /// @notice only global locker role is allowed to call in and set entered
+    /// or not entered
+    modifier onlyGlobalLockerRole() {
+        require(
+            hasRole(GLOBAL_LOCKER_ROLE, msg.sender),
+            "GlobalReentrancyLock: address missing global locker role"
+        );
+        _;
+    }
+
+    /// @notice view only function to return the last block entered
+    function lastBlockEntered() external view returns (uint88) {
+        return _lastBlockEntered;
+    }
+
+    /// @notice returns the last address that locked this contract
+    function lastSender() external view returns (address) {
+        return address(_sender);
+    }
+
+    /// @notice returns whether or not the contract is currently not entered
+    /// if true, it is possible to lock
+    function isUnlocked() external view returns (bool) {
+        return _status == _NOT_ENTERED;
+    }
+
+    /// @notice returns whether or not the contract is currently entered
+    /// if true, and locked in the same block, it is possible to unlock
+    function isLocked() external view override returns (bool) {
+        return _status == _ENTERED;
+    }
+
+    /// @notice set the status to entered
+    /// only available if not entered
+    /// callable only by global locker role
+    function lock() external override onlyGlobalLockerRole {
+        require(
+            _status == _NOT_ENTERED,
+            "GlobalReentrancyLock: system already entered"
+        );
+
+        /// cache values to save a warm SSTORE
+        uint32 blockEntered = block.number.toUint32();
```

Comment:
remove this cast entirely to save on gas as the sun will explode before this stops working